### PR TITLE
handle interrupt properly inside ClickHouseBenchmark tool

### PR DESCRIPTION
**Pull request by bharatnc**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Detailed description / Documentation draft:

Handles interrupt properly inside ClickHouseBenchmark tool. Currently the interrupt is checked but not handled properly.
This PR checks for interrupt and if it happens, sets `interrupt_received` to true (inside `tryPushQueryInteractively`) and 
handles it by exiting with exit(1). Tested with `Ctrl + C` when running the benchmark tool:
```
^CStopping launch of queries. SIGINT received.
[2]    109533 done                 echo "SELECT * FROM system.numbers LIMIT 100000 OFFSET 100000" | 
       109534 abort (core dumped)  ./clickhouse-benchmark -i 1000
``` 

Relates to: #32586

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
You get abort because `exit` does not exit immediately. It first calls global destructor. Then other threads will do something unusual due to use after free. Replace with `_exit`.

File: programs/benchmark/Benchmark.cpp [hunk 1]
```diff
@@ -330,18 +326,26 @@ class Benchmark : public Poco::Util::Application
...
                 shutdown = true;
                 break;
             }
         }
 
+        // if an interrupt is received (it's checked in tryPushQueryInteractively function call)
+        // call exit(1) so that the program actually exits on interrupts like SIGINT
+        if (interrupt_received)
+        {
+            exit(1);
```
Comment (alexey-milovidov) [inline programs/benchmark/Benchmark.cpp:346]:
```suggestion
            _exit(1);
```

**Comment by bharatnc** [comment]
> You get abort because `exit` does not exit immediately. It first calls global destructors. Then other threads will do something unusual due to use after free. Replace with `_exit`.

thanks for replacing with` _exit`. Also:

I guess I need to handle the interrupt a bit differently.
right now: exits abruptly on Ctrl +C
what we want  is: exit abruptly on pressing Ctrl+C twice or otherwise, waits and print the results before exiting.

**Comment by alexey-milovidov** [comment]
Yes, that's what we need.

File: programs/benchmark/Benchmark.cpp [hunk 2]
```diff
@@ -342,14 +346,34 @@ class Benchmark : public Poco::Util::Application
...
+            {
+            }
+            std::cout << "Second SIGINT received. Quitting immediately."
+                      << "\n";
+            _exit(1);
+        }};
+        interrupt_handler.detach();
+
+        // run the clean up in background detached so it's non-blocking
+        auto cleanup = std::thread{[&]() {
```
Comment (bharatnc) [inline programs/benchmark/Benchmark.cpp:366]: I had to move the cleanup to run in background & detached in order for the signalHandler to handle the second SIGINT  as it looks like  `pool.wait()` is blocking. However I'm noticing that with just the first Ctrl + C the shutdown is proper sometimes and often times it exits with abort core dumped. I'm not sure if there is a better way to do this.

**Comment by alexey-milovidov** [comment]
It should never trigger abort.
Calling `_exit` should be enough...

File: programs/benchmark/Benchmark.cpp [hunk 3]
```diff
@@ -342,14 +345,33 @@ class Benchmark : public Poco::Util::Application
...
+        /// block if and until signalHandler notifies and then exit.
+        auto interrupt_handler = std::thread{[&]() {
+            while (!cancel_benchmark)
+            {
+            }
+            std::cout << "Second SIGINT received. Quitting immediately."
+                      << "\n";
+            _exit(1);
+        }};
+        interrupt_handler.detach();
```
Comment (alexey-milovidov) [inline programs/benchmark/Benchmark.cpp:361]: Oops... that's extremely dangerous.

There is almost no good usage for detached threads in C++.
And there is a race condition if the program start to exit and destroy global variables and you call std::cerr.

File: programs/benchmark/Benchmark.cpp [hunk 4]
```diff
@@ -342,14 +345,33 @@ class Benchmark : public Poco::Util::Application
             }
         }
 
-        pool.wait();
-        total_watch.stop();
+        // Setup signal handler to capture additional SIGINTs (in case the user wants to quit immediately).
+        signal(SIGINT, signalHandler);
```
Comment (alexey-milovidov) [inline programs/benchmark/Benchmark.cpp:349]:
```
RETURN VALUE
       signal() returns the previous value of the signal handler, or SIG_ERR on error.  In the event of an error, errno is set to indicate the cause.
```

**Comment by alexey-milovidov** [comment]
@bharatnc I tried another solution, #33303 - please review :)

**Pull request closed without merging**